Dynamic auto scale Kinesis-Stream ingest tasks (#10985)
apache/druid: docs/development/extensions-core/kinesis-ingestion.md

|`awsAssumedRoleArn`|String|The AWS assumed role to use for additional permissions.|no|
|`awsExternalId`|String|The AWS external id to use for additional permissions.|no|
|`deaggregate`|Boolean|Whether to use the de-aggregate function of the KCL. See below for details.|no|
|`autoScalerConfig`|Object|`autoScalerConfig` to specify how to auto scale the number of Kinesis ingest tasks. See [Tasks Autoscaler Properties](#Task Autoscaler Properties) for details.|no (default == null)|

### Task Autoscaler Properties

> Note that Task AutoScaler is currently designated as experimental.

| Property | Description | Required |
| ------------- | ------------- | ------------- |
| `enableTaskAutoScaler` | Whether enable this feature or not. Set false or ignored here will disable `autoScaler` even though `autoScalerConfig` is not null| no (default == false) |
| `taskCountMax` | Maximum value of task count. Make Sure `taskCountMax >= taskCountMin`. If `taskCountMax > {numKinesisShards}`, the maximum number of reading tasks would be equal to `{numKinesisShards}` and `taskCountMax` would be ignored.  | yes |

> | `taskCountMin` | Minimum value of task count. When enable autoscaler, the value of taskCount in `IOConfig` will be ignored, and `taskCountMin` will be the number of tasks that ingestion starts going up to `taskCountMax`| yes |

Reviewer: ```suggestion
| `taskCountMin` | Minimum number of Kinesis ingestion tasks. When you enable the auto scaler, Druid ignores the value of taskCount in `IOConfig` and uses`taskCountMin` for the initial number of tasks to launch.| yes |
```

| `minTriggerScaleActionFrequencyMillis` | Minimum time interval between two scale actions | no (default == 600000) |
| `autoScalerStrategy` | The algorithm of `autoScaler`. ONLY `lagBased` is supported for now. See [Lag Based AutoScaler Strategy Related Properties](#Lag Based AutoScaler Strategy Related Properties) for details.| no (default == `lagBased`) |

### Lag Based AutoScaler Strategy Related Properties

> > Unlike the Kafka Indexing Service, Kinesis reports lag metrics measured in time milliseconds rather than message count.

Reviewer: ```suggestion
The Kinesis indexing service reports lag metrics measured in time milliseconds rather than message count which is used by Kafka.
```

| Property | Description | Required |
| ------------- | ------------- | ------------- |

> | `lagCollectionIntervalMillis` | Period of lag points collection.  | no (default == 30000) |

Reviewer: ```suggestion
| `lagCollectionIntervalMillis` | Time period between lag points collection.  | no (default == 30000) |
```
Not sure if this is the time period between collections or if it relates to some sort of time lag.

| `lagCollectionRangeMillis` | The total time window of lag collection, Use with `lagCollectionIntervalMillis`，it means that in the recent `lagCollectionRangeMillis`, collect lag metric points every `lagCollectionIntervalMillis`. | no (default == 600000) |

> | `scaleOutThreshold` | The Threshold of scale out action | no (default == 6000000) |

Reviewer: Does this mean that when the time lage reaches 6000000 (default) the autoscaler launches another task?

> | `triggerScaleOutFractionThreshold` | If `triggerScaleOutFractionThreshold` percent of lag points are higher than `scaleOutThreshold`, then do scale out action. | no (default == 0.3) |

Reviewer: Are lag points defined somewhere? Maybe we need an example of how this works together with the `scaleOutThreshold`

> | `scaleInThreshold` | The Threshold of scale in action | no (default == 1000000) |

Reviewer: same comments as for scale out

| `triggerScaleInFractionThreshold` | If `triggerScaleInFractionThreshold` percent of lag points are lower than `scaleOutThreshold`, then do scale in action. | no (default == 0.9) |

> | `scaleActionStartDelayMillis` | Number of milliseconds after supervisor starts when first check scale logic. | no (default == 300000) |

Reviewer: ```suggestion
| `scaleActionStartDelayMillis` | Number of milliseconds to delay after the supervisor starts before the first scale logic check. | no (default == 300000) |
```

> | `scaleActionPeriodMillis` | The frequency of checking whether to do scale action in millis | no (default == 60000) |

Reviewer: ```suggestion
| `scaleActionPeriodMillis` | Frequency in milliseconds to check if a scale action is triggered | no (default == 60000) |
```

> | `scaleInStep` | How many tasks to reduce at a time | no (default == 1) |

Reviewer: ```suggestion
| `scaleInStep` | Number of tasks to reduce at a time when scaling down | no (default == 1) |
```

> | `scaleOutStep` | How many tasks to add at a time | no (default == 2) |

Reviewer: ```suggestion
| `scaleOutStep` | Number of tasks to add at a time when scaling out | no (default == 2) |
```
